nixos/modules/services/web-servers/agate.nix

```diff
@@ -0,0 +1,122 @@
+        description = ''
+          Addresses to listen on, IP:PORT, if you haven't disabled forwarding
+          only set IPv4.
+        '';
+      };
+
+      contentDir = mkOption {
+        default = "/var/lib/agate/content";
+        type = types.path;
+        description = "Root of the content directory.";
+      };
+
+      certificatesDir = mkOption {
+        default = [ "/var/lib/agate/certificates" ];
+        type = types.listOf types.path;
+        description = "Root of the certificate directory.";
+      };
+
+      hostnames = mkOption {
+        default = [ ];
+        type = types.listOf types.str;
+        description = ''
+          Domain name of this Gemini server, enables checking hostname and port
+          in requests. (multiple occurences means basic vhosts)
+        '';
+      };
+
+      language = mkOption {
+        default = null;
+        type = types.nullOr types.str;
+        description = "RFC 4646 Language code for text/gemini documents.";
+      };
+
+      onlyTls_1_3 = mkOption {
+        default = false;
+        type = types.bool;
+        description = "Only use TLSv1.3 (default also allows TLSv1.2).";
+      };
+
+      extraArgs = mkOption {
+        type = types.str;
+        default = "";
+        example = "--log-ip";
+        description = "Extra arguments to use running agate.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.agate =
+      let
+        convertListArgs = key: list: concatMapStringsSep " " (value: "${key} '${value}'") list;
+        addresses = convertListArgs "--addr" cfg.addresses;
+        hostnames = convertListArgs "--hostname" cfg.hostnames;
+      in
+      {
+        description = "Agate";
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+        preStart = ''
+          if [ ! -d '${cfg.contentDir}' ]; then
+            mkdir -p '${cfg.contentDir}'
+            chown -R agate:agate '${cfg.contentDir}'
+          fi
+          if [ ! -d '${cfg.certificatesDir}' ]; then
+            mkdir -p '${cfg.certificatesDir}'
+            chown -R agate:agate '${cfg.certificatesDir}'
+          fi
+        '';
+        serviceConfig = {
+          Type = "simple";
+          PermissionsStartOnly = true;
+          ExecStart = "${cfg.package}/bin/agate ${concatStringsSep " " ([
+            "--content '${cfg.contentDir}'"
+            "--certs ${cfg.certificatesDir}"
+            addresses
+            (optionalString (cfg.hostnames != []) hostnames)
+            (optionalString (cfg.language != null) "--lang '${cfg.language}'")
+            (optionalString cfg.onlyTls_1_3 "--only-tls13")
+            cfg.extraCommandLine
+          ])}";
+          Restart = "always";
+          RestartSec = "5s";
+          User = "agate";
+          Group = "agate";
+          NoNewPrivileges = true;
+          LimitNOFILE = 65536;
+        };
+      };
+
+    environment.systemPackages = [ cfg.package ];
```

Comment:
does agate provide binaries users are expected to run? if not this might not be very useful

Reply:
other than running it there's not much use beyond --help and --version.

I guess someone might use it to generate their certificates as a dryrun and then move it to `certificatesDir`?

Reply:
I've removed it but still interested in your thoughts. I guess people can just use something else to create their certs or they can just add agate to their packages themselves

Reply:
hmm, that should be reason enough to add it to systemPackages.

Reply:
you are correct though, if it's a single action that can be easily done with other tools there's not much benefit. we don't know how complicated it is to generate certificates for agate, so really it all depends on how involved that process is. maybe other people have other views on this, but we'd only add it if three lines of `certtool` or a bit of `easy-rsa` can't do it. hope that helps.

Reply:
Looks like this would be what's required for the certs using `openssl`:

```bash
#!/usr/bin/env bash

domain="example.com"
# create private key
openssl genpkey -outform DER -out $domain/key.der -algorithm RSA -pkeyopt rsa_keygen_bits:4096

# create config file:
# the generated certificates must not be CA-capable, otherwise rustls complains
cat >openssl.conf <<EOT
[req]
default_bits = 4096
distinguished_name = req_distinguished_name
req_extensions = req_ext
prompt = no
[v3_ca]
basicConstraints = critical, CA:false
[req_distinguished_name]
commonName = $domain
[req_ext]
subjectAltName = DNS:$domain
EOT

openssl req -new -sha256 -out request.csr -key $domain/key.der -keyform DER -config openssl.conf

openssl x509 -req -sha256 -days 3650 -in request.csr -outform DER -out $domain/cert.der \
	-extensions req_ext -extfile openssl.conf -signkey $domain/key.der -keyform DER
```

https://github.com/mbrubeck/agate/blob/master/tests/data/multicert/create_certs.sh

I guess due to messing about with disabling CA-capable etc it's a bit difficult so maybe I will leave it in